queue/queue.go

```diff
@@ -0,0 +1,47 @@
+package queue
+
+import (
+	beatsqueue "github.com/elastic/beats/v7/libbeat/publisher/queue"
+
+	"github.com/elastic/elastic-agent-shipper/api"
+)
+
+// Queue is a shipper-specific wrapper around the bare libbeat queue.
+// It accepts api.Event instead of bare interface pointers like the
+// libbeat queue, and it sets opinionated defaults for the queue
+// semantics. The intention is to keep the shipper from becoming too
+// entangled with the legacy queue api, and to gradually expose more
+// features as the libbeat queue evolves and we decide what we want
+// to support in the shipper.
+type Queue struct {
+	eventQueue beatsqueue.Queue
+
+	//producer beatsqueue.Producer
+}
+
+type Metrics beatsqueue.Metrics
+
+// metricsSource is a wrapper around the libbeat queue interface, exposing only
+// the callback to query the current metrics. It is used to pass queue metrics
+// to the monitoring package.
+type MetricsSource interface {
+	Metrics() (Metrics, error)
+}
+
+func New() (*Queue, error) {
+	return &Queue{}, nil
+}
+
+func (queue *Queue) Publish(event *api.Event) {
```

Comment:
We probably want this to return `error`

Reply:
Hmm, I was thinking of this as a placeholder since it also needs to return (at least) an id of some kind and maybe other metadata we haven't decided on. But error is a fine place to start, I'll add it :-)

Reply:
Actually, on second thought, the underlying queue's `Publish` call can't return an error -- all it returns is a boolean, whether the event was added to the queue (which I believe can only be false if the queue is full during a non-blocking call). There also isn't yet anywhere to report a potential error at the call site in `shipperServer.PublishEvents` (the response only accommodates processor errors) and I'm reluctant to extend that with something we don't yet have an obvious use for... is it ok to leave this until we have something meaningful to report?

Reply:
The `PublishEvents` RPC can definitely return an error, it's in the function signature: https://github.com/elastic/elastic-agent-shipper/blob/bfd5836bb73a00243e4aea49a5e679bdd7e7ba44/server/server.go#L23

The "queue is full" case is exactly what I was thinking of. A boolean at this level works and the API can figure out what to report back to clients.

The `EventsReply` message is only returned to a client when the `error` is nil, so for processor errors the assumption is the message was enqueued but one or more of the processors failed.

Reply:
Hmm, I assumed the `PublishEvents` error return was the wrong way to return it since it's a single result for the entire request, and "couldn't add event to the queue" is an error that (as currently written) applies to individual events. If we're asked to publish 100 events and 50 are successfully added to the queue, how do we indicate that? A single catchall error and a reply that only includes the published ones?

Reply:
Right, Denis actually clarified this recently and I just forgot about it:

https://github.com/elastic/elastic-agent-shipper/blob/bfd5836bb73a00243e4aea49a5e679bdd7e7ba44/api/shipper.proto#L20-L22